// @luma.gl/shadertools/dist/es6/lib/shader-modules.js
import { assert } from '../utils';
const shaderModules = {};
let defaultShaderModules = [];
export function registerShaderModules(shaderModuleList) {
  let {
    ignoreMultipleRegistrations = false
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  for (const shaderModule of shaderModuleList) {
    registerShaderModule(shaderModule, {
      ignoreMultipleRegistrations
    });
  }
}
export function setDefaultShaderModules(modules) {
  defaultShaderModules = modules;
}
export function getShaderModule(moduleOrName) {
  if (typeof moduleOrName !== 'string') {
    const shaderModule = moduleOrName;
    assert(typeof shaderModule.name === 'string');
    registerShaderModule(shaderModule, {
      ignoreMultipleRegistrations: true
    });
    return shaderModule;
  }

  const shaderModule = shaderModules[moduleOrName];

  if (!shaderModule) {
    assert(false, "Unknown shader module ".concat(moduleOrName));
  }

  return shaderModule;
}
export function resolveModules(modules) {
  const moduleNames = modules.map(module => {
    if (typeof module !== 'string') {
      registerShaderModules([module], {
        ignoreMultipleRegistrations: true
      });
      return module.name;
    }

    return module;
  });
  return getShaderDependencies(moduleNames);
}
export function getShaderDependencies(modules) {
  modules = modules.concat(defaultShaderModules);
  const result = {};
  getDependencyGraph({
    modules,
    level: 0,
    result
  });
  return Object.keys(result).sort((a, b) => result[b] - result[a]);
}
export function getDependencyGraph(_ref) {
  let {
    modules,
    level,
    result
  } = _ref;

  if (level >= 5) {
    throw new Error('Possible loop in shader dependency graph');
  }

  for (const moduleOrName of modules) {
    const shaderModule = getShaderModule(moduleOrName);

    if (result[shaderModule.name] === undefined || result[shaderModule.name] < level) {
      result[shaderModule.name] = level;
    }
  }

  for (const moduleOrName of modules) {
    const shaderModule = getShaderModule(moduleOrName);

    if (shaderModule.dependencies) {
      getDependencyGraph({
        modules: shaderModule.dependencies,
        level: level + 1,
        result
      });
    }
  }

  return result;
}

function parseDeprecationDefinitions() {
  let deprecations = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  deprecations.forEach(def => {
    switch (def.type) {
      case 'function':
        def.regex = new RegExp("\\b".concat(def.old, "\\("));
        break;

      default:
        def.regex = new RegExp("".concat(def.type, " ").concat(def.old, ";"));
    }
  });
  return deprecations;
}

function registerShaderModule(shaderModule, _ref2) {
  let {
    ignoreMultipleRegistrations = false
  } = _ref2;
  assert(shaderModule.name, 'shader module has no name');

  if (!ignoreMultipleRegistrations && shaderModules[shaderModule.name]) {
    throw new Error("shader module ".concat(shaderModule.name, " already registered"));
  }

  shaderModules[shaderModule.name] = shaderModule;
  shaderModule.dependencies = shaderModule.dependencies || [];
  shaderModule.deprecations = parseDeprecationDefinitions(shaderModule.deprecations);
}
